apps/emqx_ds_builtin_local/src/emqx_ds_builtin_local.erl

```diff
@@ -314,7 +326,53 @@ update_iterator(DB, Iter0 = #{?tag := ?IT, ?shard := Shard, ?enc := StorageIter0
 
 -spec next(emqx_ds:db(), iterator(), pos_integer()) -> emqx_ds:next_result(iterator()).
 next(DB, Iter, N) ->
-    with_worker(do_next, [DB, Iter, N]).
+    {ok, Ref} = emqx_ds_lib:with_worker(undefined, ?MODULE, do_next, [DB, Iter, N]),
+    receive
+        #poll_reply{ref = Ref, payload = Data} ->
+            Data
+    end.
+
+-spec poll(emqx_ds:db(), emqx_ds:poll_iterators(), emqx_ds:poll_opts()) -> {ok, reference()}.
+poll(DB, Iterators, PollOpts = #{timeout := Timeout}) ->
+    %% Create a new alias, if not already provided:
+    case PollOpts of
+        #{reply_to := ReplyTo} ->
+            ok;
+        _ ->
+            ReplyTo = alias([explicit_unalias])
+    end,
+    %% Spawn a helper process that will notify the caller when the
+    %% poll times out:
+    _Completion = spawn_link(
+        fun() ->
+            wait_completion(ReplyTo, Timeout)
+        end
+    ),
+    %% Submit poll jobs:
+    lists:foreach(
+        fun({ItKey, It = #{?tag := ?IT, ?shard := Shard}}) ->
+            ShardId = {DB, Shard},
+            ReturnAddr = {ReplyTo, ItKey},
+            emqx_ds_beamformer:poll(node(), ReturnAddr, ShardId, It, PollOpts)
+        end,
+        Iterators
+    ),
+    {ok, ReplyTo}.
+
+unpack_iterator(Shard, #{?tag := ?IT, ?enc := Iterator}) ->
+    {Stream, TopicFilter, DSKey, TS} = emqx_ds_storage_layer:unpack_iterator(Shard, Iterator),
+    MsgMatcher = emqx_ds_storage_layer:message_matcher(Shard, Iterator),
+    #{
+        stream => Stream,
```

Comment:
```suggestion
unpack_iterator(Shard, #{?tag := ?IT, ?shard := Shard, ?enc := Iterator}) ->
    {InnerStream, TopicFilter, DSKey, TS} = emqx_ds_storage_layer:unpack_iterator(Shard, Iterator),
    MsgMatcher = emqx_ds_storage_layer:message_matcher(Shard, Iterator),
    #{
        stream => ?stream(Shard, InnerStream),
```
?

Reply:
This will make code more confusing. Beamformers are strictly tied to the shards, there's no point hiding of shard from them.

Reply:
Fair enough, given this API is not exposed at `emqx_ds`, there's no need to encapsulate it further.